components/brave_wallet/browser/internal/polkadot_extrinsic.rs

```diff
@@ -0,0 +1,261 @@
+        ) -> Box<CxxPolkadotDecodeUnsignedTransfer>;
+
+        fn make_chain_metadata(chain_name: &str) -> Box<CxxPolkadotChainMetadataResult>;
+
+        fn encode_unsigned_transfer_allow_death(
+            chain_metadata: &CxxPolkadotChainMetadata,
+            send_amount_bytes: &[u8; 16],
+            pubkey: &[u8; 32],
+        ) -> Vec<u8>;
+
+        fn decode_unsigned_transfer_allow_death(
+            chain_metadata: &CxxPolkadotChainMetadata,
+            mut input: &[u8],
+        ) -> Box<CxxPolkadotDecodeUnsignedTransferResult>;
+    }
+}
+
+#[macro_export]
+macro_rules! impl_result {
+    ($t:ident, $r:ident) => {
+        impl $r {
+            fn error_message(self: &$r) -> String {
+                match &self.0 {
+                    Err(e) => e.to_string(),
+                    Ok(_) => String::new(),
+                }
+            }
+
+            fn is_ok(self: &$r) -> bool {
+                self.0.is_ok()
+            }
+
+            fn unwrap(self: &mut $r) -> Box<$t> {
+                match std::mem::replace(&mut self.0, Err(Error::AlreadyUnwrapped)) {
+                    Ok(v) => Box::new(v),
+                    Err(e) => panic!("{}", e.to_string()),
+                }
+            }
+        }
+    };
+}
+
+/// Errors that can occur when parsing Polkadot runtime metadata or parsing
+/// extrinsics.
+#[derive(Clone, Debug)]
+pub enum Error {
+    /// The Result has already been unwrapped.
+    AlreadyUnwrapped,
+    /// The supplied chain name did not match our hard-coded whitelist.
+    ChainNameNotFound,
+    /// Invalid SCALE value found.
+    InvalidScale,
+    /// Invalid metadata such as the wrong pallet index or call index.
+    InvalidMetadata,
+    /// Invalid length.
+    InvalidLength,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Error::ChainNameNotFound => {
+                write!(f, "The supplied chain spec name did not match the whitelist.")
+            }
+            Error::AlreadyUnwrapped => write!(f, "Already unwrapped."),
+            Error::InvalidScale => write!(f, "Invalid SCALE-encoded bytes were found."),
+            Error::InvalidMetadata => write!(f, "Invalid chain metadata was encountered."),
+            Error::InvalidLength => write!(f, "Invalid length."),
+        }
+    }
+}
+
+struct CxxPolkadotChainMetadata {
+    balances_pallet_index: u8,
+    transfer_allow_death_call_index: u8,
+}
+
+struct CxxPolkadotChainMetadataResult(Result<CxxPolkadotChainMetadata, Error>);
+impl_result!(CxxPolkadotChainMetadata, CxxPolkadotChainMetadataResult);
+
+struct CxxPolkadotDecodeUnsignedTransferResult(Result<CxxPolkadotDecodeUnsignedTransfer, Error>);
+impl_result!(CxxPolkadotDecodeUnsignedTransfer, CxxPolkadotDecodeUnsignedTransferResult);
+
+fn make_chain_metadata(chain_name: &str) -> Box<CxxPolkadotChainMetadataResult> {
+    let metadata = match chain_name {
```

Comment:
I think this method should be in c++ and just generating CxxPolkadotChainMetadata struct.